### PR TITLE
Linearity study update 2019

### PR DESCRIPTION
**Pull request by capalmer85**
@gkrintir This is the script I used to make the PCC linearity plot.  It is flexible enough to use any tree that has nPU and some observable to check.  Please have a look and merge for your reference (and for the sake of best practices).

### CODE REVIEW

File: Studies/observablesPerPU_MC_linearity.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
+import ROOT
+import sys,os
+import numpy
+import math
+
+filename=sys.argv[1]
+#label=filename.split(".")[0]
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+
+minPU=15
+maxPU=50
+steps=5
+
+#minPU=30
+#maxPU=32
+#steps=1
+
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+maxBin=210
+#int(maxPU+math.sqrt(maxPU)*2)
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    means.append(imean)
+print means
+
+
+
+
+observables={}
+
+observables["PCC"]=[tree.nCluster,750,15000]
+observables["N. Vertices"]=[tree.nVtx,maxBin,maxBin]
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py:60]: maxBin,maxBin does it work?

File: Studies/observablesPerPU_MC_linearity.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    newtarget[mean]=pileup.Clone("newtarget"+str(mean))
+    newtarget[mean].SetTitle("target PU (mean="+str(mean)+")")
+    checkpileup[mean]=ROOT.TH1F("checkpileup","PU after re-weight (mean="+str(mean)+")",maxBin,0,maxBin)
+    last5PercentBin=0
+    last2PercentBin=0
+    last1PercentBin=0
+    for ibin in range(0,newtarget[mean].GetNbinsX()+1):
+        newtarget[mean].SetBinContent(ibin+1,ROOT.TMath.PoissonI(ibin,mean))
+        #if ibin<30:
+        #    print ibin,newtarget.GetBinContent(ibin+1)
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comments

File: Studies/observablesPerPU_MC_linearity.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    for ibin in range(0,newtarget[mean].GetNbinsX()):
+        if newtarget[mean].Integral(ibin,newtarget[mean].GetNbinsX()) > 0.01:
+            binsWithLast1Percent=ibin
+    print "mean,last1PercentBin,binsWithLast1Percent",mean,last1PercentBin,binsWithLast1Percent
+    newtarget[mean].Scale(1/newtarget[mean].Integral())
+    newtarget[mean].Draw()
+    simText.Draw("same")
+    can.Update()
+    can.SaveAs(label+"/targetPU_mean"+str(mean)+".png")
+    #raw_input()
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    newtarget[mean].Draw()
+    simText.Draw("same")
+    can.Update()
+    can.SaveAs(label+"/targetPU_mean"+str(mean)+".png")
+    #raw_input()       
+
+    reweighting[mean]=newtarget[mean].Clone("ratio")
+    reweighting[mean].Divide(pileup)
+    reweighting[mean].SetTitle("reweighting")
+    #print "Integral,bin0,1,2,3,",reweighting[mean].Integral(),reweighting[mean].GetBinContent(0),reweighting[mean].GetBinContent(1),reweighting[mean].GetBinContent(2),reweighting[mean].GetBinContent(3)
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comments

File: Studies/observablesPerPU_MC_linearity.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    for mean in means:
+        for observable in observables.keys():
+            observables[observable][5][mean].Fill(GetObs(observable),reweighting[mean].GetBinContent(int(tree.nPU)+1))
+            # am I getting the right value?
+
+        checkpileup[mean].Fill(tree.nPU,reweighting[mean].GetBinContent(int(tree.nPU)+1))
+    
+
+for mean in means:
+    #print iGraph, imean, PCCs[mean].GetMean(),checkpileup[mean].GetMeanError(), PCCs[mean].GetMeanError()
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comments

File: Studies/observablesPerPU_MC_linearity.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    newtarget[mean].SetLineColor(600)
+    newtarget[mean].SetLineWidth(2)
+    newtarget[mean].Draw("same")
+    leg.AddEntry(checkpileup[mean],"Re-weighted PU","l")
+    leg.AddEntry(newtarget[mean],"Target PU","l")
+    leg.Draw()
+    simText.Draw("same")
+    can.Update()
+    can.SaveAs(label+"/reweightingCheck_PU"+str(mean)+".png")
+    #raw_input()
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+    can.SaveAs(label+"/"+observable+str(iPlot)+"vsPU.pdf")
+    can.SaveAs(label+"/"+observable+str(iPlot)+"vsPU.root")
+    can.SaveAs(label+"/"+observable+str(iPlot)+"vsPU.C")
+    #raw_input()
+
+    iPlot=iPlot+1
+    residual=MakeResidualPlot(observables[observable][3],line)
+    residual.SetTitle(";nPU;#Delta"+observable+"_{MC-Fit}")
+    residual.Draw("APE")
+    #fitErrorGraph.Draw("same")
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        leg=ROOT.TLegend(0.35,0.45,0.87,0.85) 
+        leg.AddEntry(xOverNPU,observable+"/nPU = "+"{:.2f} #pm {:.2f}".format(fitRes.Parameter(0),fitRes.ParError(0)),"l")
+        leg.AddEntry(xOverNPU,"Constant = "+"{:.2f} #pm {:.2f}".format(fitRes.Parameter(1),fitRes.ParError(1)),"l")
+        leg.AddEntry(xOverNPU,"#chi^{2}/NDF = "+"{:.1f} / {:.1f}".format(fitRes.Chi2(),fitRes.Ndf()),"l")
+        leg.Draw("same")
+    simText.Draw("same")
+    can.Update()
+    can.SaveAs(label+"/"+observable+str(iPlot)+"vsPU.png")
+    iPlot=iPlot+1
+    #raw_input()
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py]: remove comment

File: Studies/observablesPerPU_MC_linearity.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+    for iPoint in range(100):
+        thisPU=(iPoint+0.01)/100. * realMaxPU
+        fitErrorGraph.SetPoint(iPoint,thisPU,line.Eval(thisPU))
+        fitErrorGraph.SetPointError(iPoint,0,line.IntegralError(thisPU-1,thisPU+1)/(2))
+        print iPoint,line.IntegralError(thisPU-1,thisPU+1)/2,thisPU
+
+    leg=ROOT.TLegend(0.45,0.45,0.87,0.15) 
+    leg.AddEntry(line,observable+"/nPU = "+"{:.2f} #pm {:.2f}".format(fitRes.Parameter(1),fitRes.ParError(1)),"l")
+    #leg.AddEntry(line,"noise "+"{:.4f} #pm {:.4f}".format(fitRes.Parameter(0),fitRes.ParError(0)),"l")
```
Comment (gkrintir) [inline Studies/observablesPerPU_MC_linearity.py:198]: add a comment "uncomment if noise to be added to the legend too"

**Comment by gkrintir** [comment]
thx a lot @capalmer85 

all minor comments

**Comment by capalmer85** [comment]
@gkrintir did you make any comments?  I didn't see any.

**Comment by gkrintir** [comment]
@capalmer85  right, they are inline ;D

**Comment by capalmer85** [comment]
I still don't see them.  I clicked "Files changed" to see if they were there, but I don't see any.

**Comment by gkrintir** [comment]
@capalmer85 my bad, didnt push "submit review" button

File: Studies/observablesPerPU_MC_linearity.py (hunk 4 above)
Comment (capalmer85) [reply Studies/observablesPerPU_MC_linearity.py:60]: yes.  The parts of the list are [branch, nBins, maxBin] and assumes that minBin is 0 for the plots.  For vertices, nBins and maxBin are the same so it is ok.

**Comment by capalmer85** [comment]
@gkrintir ok, I think that is everything.

**Pull request merged**